### ISSUE TITLE
Problem executing parallel requests

### ISSUE THREAD

**Issue by nbluis** [googlemaps/google-maps-services-java#254]
Hi.

I think I've identified a problem that occurs when using the library to perform parallel requests.
I'm using api with enterprise credentials and a slightly higher queryrate limit.

Well, frequently when making a request the following exception occurs.

```
java.lang.ArrayIndexOutOfBoundsException: null
	at sun.security.provider.DigestBase.engineUpdate(DigestBase.java:114)
	at sun.security.provider.SHA.implDigest(SHA.java:94)
	at sun.security.provider.DigestBase.engineDigest(DigestBase.java:181)
	at sun.security.provider.DigestBase.engineDigest(DigestBase.java:160)
	at java.security.MessageDigest$Delegate.engineDigest(MessageDigest.java:592)
	at java.security.MessageDigest.digest(MessageDigest.java:365)
	at com.sun.crypto.provider.HmacCore.engineDoFinal(HmacCore.java:201)
	at javax.crypto.Mac.doFinal(Mac.java:561)
	at javax.crypto.Mac.doFinal(Mac.java:636)
	at com.google.maps.internal.UrlSigner.getSignature(UrlSigner.java:55)
	at com.google.maps.GeoApiContext.getWithPath(GeoApiContext.java:214)
	at com.google.maps.GeoApiContext.get(GeoApiContext.java:124)
	at com.google.maps.PendingResultBase.makeRequest(PendingResultBase.java:82)
	at com.google.maps.PendingResultBase.awaitIgnoreError(PendingResultBase.java:63)

```
Debugging the code I found these snippets in [GeoApiContext](https://github.com/googlemaps/google-maps-services-java/blob/master/src/main/java/com/google/maps/GeoApiContext.java#L257) and [UrlSigner](https://github.com/googlemaps/google-maps-services-java/blob/master/src/main/java/com/google/maps/internal/UrlSigner.java#L47).

The implementation of URLSigner that is instantiated by GeoApiContext uses only one instance of javax.crypto.Mac. 

But according to [this](http://stackoverflow.com/questions/31898802/is-macdofinal-thread-safe-in-java)  javax.crypto.Mac keeps internal state to perform encryption and is not thread safe.

It would be necessary to create a new instance (or clone a fixed one) to each new signature to avoid concurrency problems.

I think I can try to send a pull request to fix this.

What do you think about ?

**Comment by nbluis** [comment]
Any thoughts ?

**Comment by domesticmouse** [comment]
I'm unsure why I closed this 7 days ago. I believe this is now fixed with v0.1.20.
